File: 567. Permutation in String/solution.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
       int n1 = s1.length(), n2 = s2.length(), l = 0, r = 0;
       int[] count = new int[26];

       for (char c : s1.toCharArray()) {
            count[c - 'a']++;
        }

        while (r < n2) {
            count[s2.charAt(r) - 'a']--;
            n1--;

            while (count[s2.charAt(r) - 'a'] < 0) {
                count[s2.charAt(l++) - 'a']++;
                n1++;
            }

            if (n1 == 0)
                return true;
            r++;
        }

        return false;
    }
}

File: 567. Permutation in String/solution.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        n1, n2, l, r = len(s1), len(s2), 0, 0
        count = [0] * 26

        for c in s1:
            count[ord(c) - ord('a')] += 1

        while r < n2:
            count[ord(s2[r]) - ord('a')] -= 1
            n1 -= 1

            while count[ord(s2[r]) - ord('a')] < 0:
                count[ord(s2[l]) - ord('a')] += 1
                l += 1
                n1 += 1

            if n1 == 0:
                return True
            r += 1

        return False
        